src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs

```diff
@@ -0,0 +1,83 @@
+    /// </summary>
+    internal class TokenReadingResult: ValidationResult
+    {
+        private Exception? _exception;
+        private SecurityToken? _securityToken;
+
+        /// <summary>
+        /// Creates an instance of <see cref="TokenReadingResult"/>.
+        /// </summary>
+        /// <paramref name="tokenInput"/> is the string from which the <see cref="SecurityToken"/> was created.
+        /// <paramref name="securityToken"/> is the <see cref="SecurityToken"/> that was created.
+        public TokenReadingResult(SecurityToken securityToken, string tokenInput)
+            : base(ValidationFailureType.ValidationSucceeded)
+        {
+            IsValid = true;
+            TokenInput = tokenInput;
+            _securityToken = securityToken;
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="TokenReadingResult"/>
+        /// </summary>
+        /// <paramref name="tokenInput"/> is the string that failed to create a <see cref="SecurityToken"/>.
+        /// <paramref name="validationFailure"/> is the <see cref="ValidationFailureType"/> that occurred during reading.
+        /// <paramref name="exceptionDetail"/> is the <see cref="ExceptionDetail"/> that occurred during reading.
+        public TokenReadingResult(string? tokenInput, ValidationFailureType validationFailure, ExceptionDetail exceptionDetail)
+            : base(validationFailure, exceptionDetail)
+        {
+            TokenInput = tokenInput;
+            IsValid = false;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="Exception"/> that occurred during reading.
+        /// </summary>
+        public override Exception? Exception
+        {
+            get
+            {
+                if (_exception != null || ExceptionDetail == null)
+                    return _exception;
+
+                HasValidOrExceptionWasRead = true;
+                _exception = ExceptionDetail.GetException();
+
+                if (_exception is SecurityTokenException securityTokenException)
+                {
+                    securityTokenException.Source = "Microsoft.IdentityModel.Tokens";
+                    securityTokenException.ExceptionDetail = ExceptionDetail;
+                }
+
+                return _exception;
+            }
+        }
+
+        /// <summary>
+        /// Gets the string from which the <see cref="SecurityToken"/> was read.
+        /// </summary>
+        public string? TokenInput { get; }
+
+        /// <summary>
+        /// Gets the <see cref="SecurityToken"/> that was read.
+        /// </summary>
+        public SecurityToken SecurityToken {
+            get
+            {
+                if (_securityToken is null)
+                    throw new InvalidOperationException("Attempted to retrieve the SecurityToken from a failed TokenReading result.");
```

Comment:
generally it is an anti pattern to throw exceptions from property getters

Reply:
Changed to method and documented the exception.